File: src/main/router.ts
// Routing logic for the LLM <-> agent-harness coordinator. Pure + unit-testable.
//
// Decides whether a user message should go to the regular conversational LLM or
// the agent harness. Explicit phrasing wins; otherwise an agentic-intent
// heuristic routes coding/tool/file/system tasks to the harness.

export type Target = 'llm' | 'harness';

// Explicit "use the agent/harness" (or the opposite) phrasing.
const EXPLICIT_HARNESS = /\b(use|using|ask|via|with|through)\s+(the\s+)?(agent|harness|coder?|codex|claude\s*code)\b|^\s*(agent|harness)[,:]/i;
const EXPLICIT_LLM = /\b(just\s+(chat|talk|answer)|no\s+(agent|harness|code)|don'?t\s+use\s+the\s+(agent|harness))\b/i;

// Agentic-intent keywords (coding / files / system) AND tool/real-time intent
// (live data or device actions the harness can do via tools, which a plain chat
// LLM cannot). Matched anywhere in the message. Deliberately broad: for a voice
// assistant, when a request plausibly needs a tool we prefer the tool-capable
// harness over the chat LLM — the old narrow list is what let clear tool
// requests ("what time is it", "set a timer", "will it rain tomorrow") fall
// through to the direct LLM. Ambiguous VERBS (open/play/send/…) live in ACTION
// below (start-anchored) so conversational filler ("in order to", "note that")
// doesn't trip them.
const AGENTIC = new RegExp(
  '\\b(' + [
    // coding / files / system
    'code', 'coding', 'refactor', 'refactoring', 'debug', 'debugging', 'bug', 'fix',
    'implement', 'implementation', 'function', 'class', 'method', 'variable',
    'file', 'files', 'directory', 'folder', 'repo', 'repository', 'commit', 'branch',
    'pull request', 'merge', 'diff', 'git',
    'run', 'execute', 'build', 'compile', 'deploy', 'install', 'uninstall', 'script', 'command',
    'terminal', 'shell', 'test', 'tests', 'lint', 'package', 'dependency',
    'api', 'endpoint', 'database', 'query', 'sql', 'server', 'docker',
    'edit', 'rename', 'delete', 'create a', 'write a', 'add a',
    // weather / environment (live)
    'weather', 'forecast', 'temperature', 'humidity', 'raining', 'rain', 'snow',
    'sunny', 'cloudy', 'windy', 'storm', 'umbrella', 'sunrise', 'sunset',
    'uv index', 'air quality', 'pollen',
    // time / date (live) — more phrasing variants so a paraphrased "what's the
    // current time" or "do you know what time it is" reliably routes to the
    // harness. The old list missed "do you know" / "can you tell me" prefixes,
    // so casual phrasings leaked to the chat LLM and got hallucinated answers.
    'what time', 'time is it', 'time it is', 'do you know what time', 'tell me the time',
    'what day', 'what.s the date', 'todays date', "today's date", 'date today',
    'current time', 'current date',
    // news / finance / sports (live)
    'news', 'headlines', 'stock', 'stocks', 'shares', 'market', 'crypto',
    'bitcoin', 'ethereum', 'price of', 'how much is', 'exchange rate', 'currency',
    'score', 'scores', 'who won', 'standings', 'who is winning', 'latest score',
    // search / web / research
    'search', 'search for', 'look up', 'lookup', 'google', 'bing', 'wikipedia',
    'browse', 'website', 'on the internet',
    // navigation / places (live)
    'directions', 'navigate', 'route to', 'nearest', 'nearby', 'near me',
    'traffic', 'how far', 'how long to get',
    // device / system actions
    'volume', 'brightness', 'mute', 'flashlight', 'wifi', 'bluetooth',
    'battery', 'screenshot', 'screen', 'what.s on my',
    // comms / productivity
    'email', 'inbox', 'whatsapp', 'slack', 'calendar', 'meeting', 'appointment',
    'schedule a', 'remind me', 'reminder', 'set a timer', 'set an alarm',
    'alarm', 'timer', 'shopping list', 'add to my',
    // conversions / utilities / commerce
    'convert', 'how many', 'translate', 'translation', 'calculate',
    'book a', 'place an order', 'reserve a',
  ].join('|') + ')\\b',
  'i',
);

// Strong real-time signals: when present, the answer depends on the live world,
// so it needs tools -> harness. Kept tight (no bare "today"/"tonight") so casual
// pleasantries like "how are you today" aren't misrouted. Added a few more
// variants (right now / currently / latest) and an explicit "what is the
// weather/temperature/forecast/time" so the common phrasings never fall through.
const REALTIME =
  /\b(right now|currently|the latest|up[- ]?to[- ]?date|near me|nearby|around here|in my area|my area|local events?|this (week|weekend|month|year)|what time|what'?s the time|what day|what'?s the date|what is the (weather|time|forecast|date|temperature|score|price)|events? (today|tonight|tomorrow|yesterday|last night|near me|in my area)|fireworks? (show|shows|event|events|happened|near|tonight|tomorrow|yesterday|last night)|happened (yesterday|last night))\b/i;

// Imperative device/tool actions at the START of the message -> harness. Limited
// to verbs that imply DOING something (not "tell/explain/describe/what/how",
// which are conversational), and start-anchored so they don't match mid-sentence
// filler. Added a few more common action verbs ("set", "start", "stop", "go to")
// so a paraphrased "go to wikipedia" or "set brightness to 50" reliably routes.
const ACTION =
  /^\s*(open|launch|play|pause|resume|skip|mute|unmute|turn|set|send|call|text|email|remind|schedule|book|order|buy|reserve|navigate|download|install|update|upgrade|enable|disable|check|find|search|look up|show me|get me|pull up|bring up|take a|start|stop|go to|switch|toggle|change|adjust|raise|lower|increase|decrease)\b/i;

// Screen-share vision detail. The OpenAI-compatible `image_url.detail` controls
// how hard the vision model works: "high" tiles the image into 512px tiles (many
// tokens, slow TTFT) while "low" is a single ~512px low-res pass (flat, fast). A
// general "what's on my screen / what am I looking at" glance doesn't need fine
// detail, so it goes "low" for a much faster reply; anything that implies READING
// fine content (text, code, an error) keeps "high" so legibility isn't lost. This
// is the main lever on the "every turn is slow while screen sharing" delay.
const VISION_GLANCE =
  /\b(what'?s on (my|the) (screen|display|monitor)|what am i (looking at|seeing|on)|what (app|window|program|tab|page|site)|which (app|window|program|tab)|what'?s this|what is this|what do you see|describe (my|the|this) (screen|display|page|window)|give me (a|an) (overview|summary) of (my|the) screen)\b/i;

export function visionDetailFor(message: string): 'low' | 'high' {
  return VISION_GLANCE.test(message || '') ? 'low' : 'high';
}

export interface RouteConfig {
  mode: 'auto' | 'llm' | 'harness';
  hasLlm: boolean;       // a conversational LLM endpoint is configured
  hasHarness: boolean;   // an agent harness endpoint is configured
  lastTarget?: Target | null; // which target handled the previous turn (for stickiness)
  lastWasQuestion?: boolean;  // the previous reply ended with a question (awaiting an answer)
}

// A short reply with no fresh intent is treated as a continuation of the current
// turn (e.g. answering the harness's "where are you?" with "Austin, Texas").
function isContinuation(text: string): boolean {
  const words = text.trim().split(/\s+/).filter(Boolean);
  return words.length > 0 && words.length <= 8;
}

/**
 * Choose a target for `message` given availability + mode.
 * Falls back to whichever is configured if the preferred one isn't.
 */
export function route(message: string, cfg: RouteConfig): Target {
  // Honor a hard mode override (still falling back if that one isn't configured).
  if (cfg.mode === 'llm') return cfg.hasLlm ? 'llm' : 'harness';
  if (cfg.mode === 'harness') return cfg.hasHarness ? 'harness' : 'llm';

  // auto: only one configured -> use it.
  if (cfg.hasHarness && !cfg.hasLlm) return 'harness';
  if (cfg.hasLlm && !cfg.hasHarness) return 'llm';
  if (!cfg.hasLlm && !cfg.hasHarness) return 'llm';

  // Both configured: decide by intent.
  const text = message || '';
  if (EXPLICIT_LLM.test(text)) return 'llm';
  if (EXPLICIT_HARNESS.test(text)) return 'harness';
  if (AGENTIC.test(text) || REALTIME.test(text) || ACTION.test(text)) return 'harness';
  // Stickiness: if the agent harness handled the previous turn, keep this turn on
  // the harness when it's a continuation — either a short follow-up OR an answer
  // to a question the harness just asked (e.g. it asked "where are you?" and the
  // user replies with their location). This keeps the multi-turn task on one
  // target with full context instead of dropping the answer to the LLM. Explicit
  // "just chat" above already escapes this.
  if (cfg.lastTarget === 'harness' && (cfg.lastWasQuestion || isContinuation(text))) return 'harness';
  return 'llm';
}

File: apps/aria/src/router.rs
//! Mixture-mode router: per-query choice between the agent harness (tools)
//! and an optional direct LLM (fast tool-less chat). Both share one session
//! history, and a harness follow-up question makes the next reply sticky so
//! handoffs complete ("where are you?" → "Miami" must go back to the harness).

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Route {
    Harness,
    Direct,
}

/// Queries that need live data, actions, or device access — never answerable
/// from a tool-less model without hallucinating (the v2 failure mode).
const TOOL_INTENTS: &[&str] = &[
    // live data
    "weather", "temperature", "forecast", "news", "stock", "price of", "score",
    "today", "tonight", "tomorrow", "right now", "currently", "latest", "current",
    "what time", "date today",
    // web
    "search", "look up", "google", "browse", "website", "link me", "find me",
    // actions & device
    "open ", "run ", "execute", "install", "download", "launch", "play ",
    "file", "folder", "directory", "screenshot", "screen share", "my screen",
    "clipboard", "remind", "timer", "alarm", "schedule", "calendar", "email",
    "send ", "message ", "call ",
    // personal/contextual data a bare LLM can't know
    "my location", "near me", "where am i",
];

pub struct Router {
    /// "auto" | "direct" | "harness"
    mode: String,
    has_direct: bool,
    /// Set when the harness's last reply asked the user something — the
    /// user's next message returns to the harness even if it looks simple.
    sticky_harness: bool,
}

impl Router {
    pub fn new(mode: &str, has_direct: bool) -> Self {
        Self { mode: mode.to_string(), has_direct, sticky_harness: false }
    }

    pub fn route(&self, text: &str) -> Route {
        if !self.has_direct || self.mode == "harness" {
            return Route::Harness;
        }
        if self.mode == "direct" {
            return Route::Direct;
        }
        // auto
        if self.sticky_harness {
            return Route::Harness;
        }
        let t = text.to_lowercase();
        if TOOL_INTENTS.iter().any(|k| t.contains(k)) {
            Route::Harness
        } else {
            Route::Direct
        }
    }

    /// Feed back each assistant reply so handoff continuations stick.
    pub fn observe_reply(&mut self, route: Route, reply: &str) {
        self.sticky_harness = route == Route::Harness && reply.trim_end().ends_with('?');
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn no_direct_configured_means_harness_always() {
        let r = Router::new("auto", false);
        assert_eq!(r.route("hello there"), Route::Harness);
    }

    #[test]
    fn auto_routes_tool_intents_to_harness() {
        let r = Router::new("auto", true);
        for q in [
            "what's the weather for my location",
            "search for rust audio crates",
            "open the downloads folder",
            "remind me at 6pm",
            "what's the latest news",
        ] {
            assert_eq!(r.route(q), Route::Harness, "{q}");
        }
        for q in ["hello how are you", "tell me a joke", "explain quicksort simply"] {
            assert_eq!(r.route(q), Route::Direct, "{q}");
        }
    }

    #[test]
    fn harness_question_makes_next_turn_sticky() {
        let mut r = Router::new("auto", true);
        assert_eq!(r.route("what's the weather"), Route::Harness);
        r.observe_reply(Route::Harness, "Sure — where are you located?");
        // A bare location looks "simple" but must return to the harness.
        assert_eq!(r.route("Miami, Florida"), Route::Harness);
        r.observe_reply(Route::Harness, "It's 91 degrees and sunny in Miami.");
        assert_eq!(r.route("thanks! tell me a joke"), Route::Direct);
    }

    #[test]
    fn forced_modes_override_intent() {
        let r = Router::new("direct", true);
        assert_eq!(r.route("what's the weather"), Route::Direct);
        let r = Router::new("harness", true);
        assert_eq!(r.route("tell me a joke"), Route::Harness);
    }
}
